### ISSUE TITLE
sendfile: VLA buffer can be too big for stack

### ISSUE THREAD

**Issue by ghost** [svoboda18/libnt#1]
this causes [svoboda18/magiskboot](https://github.com/svoboda18/magiskboot) to fail on repacking some boot images, and stack overflow occurs.

i think we can use a smaller fixed-size buffer (for example 8192 bytes or sth) instead.

**Comment by ghost** [comment]
btw, i also think the current sendfile implementation isn't complete and it can probably be improved, here is my version (i am a noob and may not be 100% correct):

````diff
diff --git a/libnt.c b/libnt.c
index 21408ac..b37d3e3 100644
--- a/libnt.c
+++ b/libnt.c
@@ -521,42 +521,76 @@ FILE *__NT_DCL fdopen2(int fd, const char *mode) {
     return orig_fdopen(fd, p_mode.mode);
 }
 
+#undef  MIN
+#define MIN(a, b)  (((a) < (b)) ? (a) : (b))
+
+#define _SENDFILE_BUFFSIZE	8192
+
 ssize_t __NT_DCL sendfile(int out_fd, int in_fd, off_t *offset, size_t count) {
     ssize_t bytes = 0;
+    char buf[_SENDFILE_BUFFSIZE];
+    off_t orig_pos;
 
-    assert(out_fd > -1 && in_fd > -1);
+    if (out_fd < 0 || in_fd < 0) {
+        errno = EBADF;
+        return -1;
+    }
+
+    if (count < 0) {
+        errno = EINVAL;
+        return -1;
+    } else if (count == 0)
+        return 0;
 
     if (offset) {
+        orig_pos = lseek(in_fd, 0, SEEK_CUR);
+        NT_DEBUG("original file pos %lld", orig_pos);
+        if (orig_pos == (off_t)-1)
+            return -1;
         off_t pos = lseek(in_fd, *offset, SEEK_SET);
         NT_DEBUG("changed pos to %lld", pos);
-        if (pos == (off_t)-1 || pos != *offset)
+        if (pos == (off_t)-1)
             return -1;
     }
 
     while (count) {
-        char buf[count];
-        NT_DEBUG("need %llu", count);
-        ssize_t rd = read(in_fd, buf, count);
+        size_t need = MIN(count, _SENDFILE_BUFFSIZE);
+        NT_DEBUG("need %llu", need);
+        ssize_t rd = read(in_fd, buf, need);
         NT_DEBUG("read %lld", rd);
-        if (rd < 0) {
-            bytes = -1;
+        if (rd < 0 || rd == 0) // err or EOF
             break;
-        }
+        if (offset)
+            *offset += rd;
         ssize_t we = write(out_fd, buf, rd);
         NT_DEBUG("write %lld", we);
-        if (we < 0 || rd != we) {
-            bytes = -1;
+        assert(we != 0); // unspecified behavior
+        if (we < 0)
             break;
-        }
         bytes += we;
-        if (offset)
-            *offset += rd;
-        if ((size_t)rd < count) // short read
+        while (we < rd) { // writing got interrupted, retry
+            ssize_t remaining = rd - we;
+            NT_DEBUG("remaining %lld", remaining);
+            ssize_t wr = write(out_fd, buf + we, remaining);
+            NT_DEBUG("write remaining %lld", wr);
+            assert(wr != 0);
+            if (wr < 0)
+                goto exit_loop;
+            we += wr;
+            bytes += wr;
+        }
+        if ((size_t)rd < need) // short read
             break;
         count -= rd;
     }
-    if (offset)
+exit_loop:
+    if (!bytes) // written nothing, zero count is handled above
+        bytes = -1;
+    if (offset) {
         NT_DEBUG("offset %lld", *offset);
+        assert(lseek(in_fd, orig_pos, SEEK_SET) == orig_pos);
+        NT_DEBUG("restore file pos %lld", orig_pos);
+    }
     NT_DEBUG("bytes %lld", bytes);
     return bytes;
 }
````

**Comment by ghost** [comment]
> this causes [svoboda18/magiskboot](https://github.com/svoboda18/magiskboot) to fail on repacking some boot images, and stack overflow occurs.

crash log (i can see these when building with `DEBUG` enabled):

![image](https://user-images.githubusercontent.com/103345643/227764749-1b6bcbc2-6010-4b8b-8972-c873f847eb2b.png)

the exception code indicates there is a stack overflow occurred, and it fails in the _chkstk_ms function.

![image](https://user-images.githubusercontent.com/103345643/227764882-e75cb494-bde8-4c57-8c28-d2229fea9fd0.png)

after going through some previous stack frames, i can see these:

![image](https://user-images.githubusercontent.com/103345643/227764601-287ef0e4-725c-48db-8bc3-44c23d573f5b.png)

**Comment by ghost** [comment]
you can reproduce by giving `count` a big number, and call sendfile on whatever files i guess

for example, create a big file (not sure how big, i chose 16MiB+) `test.in` and an empty file `test.out`:

````c
#include "libnt.h"

int main() {
       	int in_fd, out_fd;

       	in_fd = open("test.in", O_RDONLY);
       	out_fd = open("test.out", O_WRONLY);

       	sendfile(out_fd, in_fd, NULL, 1024 * 1024 * 16);

       	close(in_fd);
       	close(out_fd);

       	return 0;
}
````

this triggers the bug, and if you reduce the `count` to a small number, nothing will be wrong.

**Comment by ghost** [comment]
after changing the VLA buffer to a fixed small buffer like i did in the above patch, the error is gone.

**Comment by ghost** [comment]
> i think we can use a smaller fixed-size buffer (for example 8192 bytes or sth) instead.

or maybe we can also use malloc() to allocate that big buffer, so we won't need to worry about overflowing the stack by a big VLA.

**Comment by svoboda18** [comment]
this is fixed on f3027ff. Thanks for your contribution.
